core/textinput/src/textinput/Editor.cpp

```diff
@@ -90,40 +90,41 @@ namespace textinput {
     fSearch.clear();
     fUndoBuf.clear();
     CancelSpecialInputMode(R);
-    if (fReplayHistEntry != (size_t) -1) {
+    if (fReplayHistEntry != static_cast<size_t>(-1)) {
       --fReplayHistEntry; // intentional overflow to -1
       fContext->SetLine(fContext->GetHistory()->GetLine(fReplayHistEntry));
     }
     return R;
   }
 
   void
-  Editor::SetReverseHistSearchPrompt(Range& RDisplay) {
-    std::string P("[bkw'");
-    SetEditorPrompt(Text(P + fSearch + "'] "));
+  Editor::SetHistSearchModePrompt(Range& RDisplay) {
+    const std::string direction(fMode == kHistFwdSearchMode ? "fwd" : fMode == kHistRevSearchMode ? "bkw" : "NA");
+    SetEditorPrompt(Text("[" + direction + "'" + fSearch + "'] "));
     RDisplay.ExtendPromptUpdate(Range::kUpdateEditorPrompt);
   }
 
   bool
   Editor::UpdateHistSearch(EditorRange& R) {
     History* Hist = fContext->GetHistory();
     Text& Line = fContext->GetLine();
-    size_t NewHistEntry = (size_t) -1;
+    size_t NewHistEntry = static_cast<size_t>(-1);
     if (fSearch.empty()) {
       NewHistEntry = 0;
     } else {
       size_t startAt = fCurHistEntry;
-      if (startAt == (size_t) -1) {
+      if (startAt == static_cast<size_t>(-1)) {
         startAt = 0;
       }
-      for (size_t i = startAt, n = Hist->GetSize(); i < n; ++i) {
+      const bool forward = fMode == kHistFwdSearchMode;
+      for (long i = startAt ; forward ? i >= 0 : i < static_cast<long>(Hist->GetSize()); forward ? --i : ++i) {
```

Comment:
The suggestion below is not tested, but it might be slightly more efficient.
```suggestion
      const ssize_t stopAt = (fMode == kHistFwdSearchMode ? -1 : static_cast<ssize_t>(Hist->GetSize());
      const ssize_t step = (fMode == kHistFwdSearchMode ? -1 : 1);
      for (ssize_t i = startAt; i != stopAt; i += step) {
```